### DISCUSSION TITLE
Guide on plugin development

### DISCUSSION THREAD

**Discussion by p-kraszewski** [strongswan/strongswan#1575] [General]
Hello!

Is there an official guide on how to _develop_ your own plugins? Like on plugin life cycle, ways to (de)register, live update config, etc?

I creates XAUTH plugin by inspecting some pre-existing, standard plugins, but I had to copy plenty of code without any explanation.

One thing that bothers me the most: are XAUTH's `process` and `initiate` callbackes called in multithreaded fashion or are they some way sequenced?

Best regards,

**Comment by tobiasbrunner** [comment]
> Is there an official guide on how to _develop_ your own plugins?

Other than the [API docs](https://www.strongswan.org/apidoc/) (see e.g. [the `xauth_method_t` interface](https://www.strongswan.org/apidoc/structxauth__method__t.html)), no. And, of course, the source code as there are a lot of existing plugins.

> I creates XAUTH plugin

Why?!? IKEv1 is dead and should never be used (or even considered) anymore.

> One thing that bothers me the most: are XAUTH's `process` and `initiate` callbackes called in multithreaded fashion or are they some way sequenced?

Each IKE_SA will have their own instance of the `xauth_method_t` implementation your plugin provides. So unless these instances access some shared objects, multi-threading isn't an issue.

**Comment by p-kraszewski** [comment]
> Why?!?

I'm fixing pre-existing 3-rd party code. If you have some hint on writing a modern form of authentication plugin that sees *plaintext* password, I'm all ears.

> So unless these instances access some shared objects, multi-threading isn't an issue.

I *do* need some shared data, that's why I'm asking at all.

**Reply by tobiasbrunner** [reply]
> If you have some hint on writing a modern form of authentication plugin that sees _plaintext_ password, I'm all ears.

There is the [_eap-gtc_ plugin](https://docs.strongswan.org/docs/5.9/plugins/eap-gtc.html).

**Reply by p-kraszewski** [reply]
Phew... Ok - fortunately, it *is* using `eap-gtc → xauth-pam` path. I'm replacing `xauth-pam` part with my implementation. You scared the sh*t out of me with IKEv1.

And for further users: due to multithreading you confirmed, `eap-gtc → xauth-pam → pam-python` will blow in your face under the slightest hint of higher load. That's why my client hired me to fix the mess the previous team implemented.

**Reply by tobiasbrunner** [reply]
I see, then that's OK :)

The _xauth-pam_ plugin creates a new handle (`pam_start()`) for each authentication, which should be thread-safe at this level (would be weird otherwise). So that would indicate a problem in the backend (e.g. whatever `pam-python` does).

**Reply by p-kraszewski** [reply]
I didn't investigate that much deeper, but `pam-python` stack dumps clearly indicated MT woes.

Thanks for all the info, best regards.
